tracetypes/org.eclipse.tracecompass.incubator.opentracing.core/src/org/eclipse/tracecompass/incubator/opentracing/core/analysis/callstack/AsincCallStackStateProvider.java

```diff
@@ -0,0 +1,209 @@
+        fStackDepthMap = new HashMap<>();
+        fPrevEvent = new TreeMap<>();
+    }
+
+    @Override
+    public int getVersion() {
+        return 3;
+    }
+
+    @Override
+    public @NonNull ITmfStateProvider getNewInstance() {
+        return new AsincCallStackStateProvider(getTrace());
+    }
+
+    @Override
+    protected void eventHandle(@NonNull ITmfEvent event) {
+        ITmfStateSystemBuilder ss = getStateSystemBuilder();
+        if (ss == null) {
+            return;
+        }
+        handleSpan(event,ss);
+    }
+
+    private void handleSpan(ITmfEvent event, ITmfStateSystemBuilder ss) {
+        long timestamp = event.getTimestamp().toNanos();
+        if (timestamp == getTrace().getStartTime().toNanos()) {
+            timestamp++;
+        }
+        Long duration = event.getContent().getFieldValue(Long.class, IOpenTracingConstants.DURATION);
+        if (duration == null) {
+            return;
+        }
+        while (!(fPrevEvent.isEmpty()) && (fPrevEvent.firstKey() < timestamp)) {
+            long prevTime = fPrevEvent.firstKey();
+            List<Integer> quarks = Objects.requireNonNull(fPrevEvent.get(fPrevEvent.firstKey()));
+            ss.modifySpanAttribute(prevTime, (Object) null, quarks.get(0), quarks.get(1));
+            fPrevEvent.remove(fPrevEvent.firstKey());
+        }
+
+        String processName = event.getContent().getFieldValue(String.class, IOpenTracingConstants.TRACE_ID);
+
+        int processId = getProcessId(event);
+        if (processName == null) {
+            processName = (processId == UNKNOWN_PID) ? UNKNOWN : Integer.toString(processId);
+        }
+
+        int pq = ss.getQuarkAbsoluteAndAdd(PROCESSES, processName);
+        ss.updateOngoingState(TmfStateValue.newValueInt(processId), 1);
+
+        String Opname = String.valueOf(TmfTraceUtils.resolveAspectOfNameForEvent(event.getTrace(), "Name", event)); //$NON-NLS-1$
+        String spanId = event.getContent().getFieldValue(String.class, IOpenTracingConstants.SPAN_ID);
+        String parentId = event.getContent().getFieldValue(String.class, IOpenTracingConstants.REFERENCES + "/CHILD_OF"); //$NON-NLS-1$
+
+        int callStackQuark = ss.getQuarkRelativeAndAdd(pq, InstrumentedCallStackAnalysis.CALL_STACK);
+        int stackDepth = getStackDepth(parentId);
+
+        Object functionEntryName = functionEntry(spanId, parentId, Opname);
+        stackDepth++;
+
+        int spanQuark = ss.getQuarkRelativeAndAdd(callStackQuark, String.valueOf(stackDepth));
+
+        ss.modifySpanAttribute(timestamp, functionEntryName, fStateQuark, spanQuark);
+
+        fStackDepthMap.put(spanId, stackDepth);
+
+        List<Integer> quarksList = Arrays.asList(fStateQuark, spanQuark);
+        fPrevEvent.put(timestamp + duration, quarksList);
+        fStateQuark++;
+
+    }
+
+    protected @Nullable String getProcessName(ITmfEvent event) {
+
+        Long fieldValue = event.getContent().getFieldValue(Long.class, "trace_id");
```

Comment:
Missing comment about non-exeternalized string `//$NON-NLS-<n>$`. Other strings in this file have the same issue.